File: frontend/src/types/camera-config.ts
// frontend/src/types/camera-config.ts
export interface CameraOption {
  camera_id: string;
}

export interface SeatModelOption {
  seat_model_id: string;
  display_name: string;
  cameras: CameraOption[];
}

export interface SeatModel {
  id: string;
  seat_model_id: string;
  display_name: string;
  yolo_model_version_id: string | null;
  projector_model_version_id: string | null;
  whitening_matrix_model_version_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface CameraConfig {
  id: string;
  camera_id: string;
  seat_model_id: string;
  patchcore_model_version_id: string | null;
  filter_classifier_model_version_id: string | null;
  normalizer_model_version_id: string | null;
  detection_confidence: number;
  patchcore_image_size: number;
  patchcore_threshold: number;
  regions: CameraRegionConfig[];
  created_at: string;
  updated_at: string;
}

export interface CameraRegionConfig {
  id?: string;
  region_id: string;
  box?: [number, number, number, number] | null;
  patchcore_model_version_id: string | null;
  enabled?: boolean;
  sort_order?: number;
  patchcore?: Record<string, object | undefined> | null;
}

export interface CameraConfigFormData {
  camera_id: string;
  patchcore_model_version_id: string | null;
  filter_classifier_model_version_id: string | null;
  normalizer_model_version_id: string | null;
  detection_confidence: number;
  patchcore_image_size: number;
  patchcore_threshold: number;
  regions: CameraRegionConfig[];
}

export interface SeatModelFormData {
  seat_model_id: string;
  display_name: string;
  yolo_model_version_id: string | null;
  projector_model_version_id: string | null;
  whitening_matrix_model_version_id: string | null;
}

/** 模型选项（下拉框用） */
export interface ModelOption {
  model_id: string;
  model_name: string;
  version: string;
  model_type: string;
  artifact_path: string;
}

/** 注册外部模型 */
export interface ModelRegisterData {
  model_name: string;
  version: string;
  model_type: string;
  artifact_path: string;
}

File: backend/app/schemas/camera_config.py
# backend/app/schemas/camera_config.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


# ── Seat Model ──

class SeatModelCreate(BaseModel):
    seat_model_id: str = Field(..., max_length=128, description="座椅型号标识符")
    display_name: str = Field(..., max_length=256, description="显示名称")
    yolo_model_version_id: str | None = Field(default=None, description="全局 YOLO 检测模型版本 ID")
    projector_model_version_id: str | None = Field(default=None, description="全局 EmbeddingProjector 模型版本 ID")
    whitening_matrix_model_version_id: str | None = Field(default=None, description="全局 WhiteningTransform 模型版本 ID")


class SeatModelUpdate(BaseModel):
    seat_model_id: str | None = Field(default=None, max_length=128)
    display_name: str | None = Field(default=None, max_length=256)
    yolo_model_version_id: str | None = Field(default=None, description="全局 YOLO 检测模型版本 ID")
    projector_model_version_id: str | None = Field(default=None, description="全局 EmbeddingProjector 模型版本 ID")
    whitening_matrix_model_version_id: str | None = Field(default=None, description="全局 WhiteningTransform 模型版本 ID")


class SeatModelResponse(BaseModel):
    id: str
    seat_model_id: str
    display_name: str
    yolo_model_version_id: str | None = None
    projector_model_version_id: str | None = None
    whitening_matrix_model_version_id: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SeatModelWithCameras(SeatModelResponse):
    cameras: list[CameraConfigResponse] = []


# ── Camera Config ──

class CameraRegionCreate(BaseModel):
    region_id: str = Field(..., max_length=64, description="ROI 区域 ID")
    patchcore_model_version_id: str = Field(..., description="区域 PatchCore 模型版本 ID")


class CameraRegionResponse(BaseModel):
    id: str
    region_id: str
    patchcore_model_version_id: str
    box: list[float] | None = None
    enabled: bool = True
    sort_order: int = 0
    patchcore: dict[str, object] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CameraConfigCreate(BaseModel):
    camera_id: str = Field(..., max_length=128, description="相机标识符")
    efficientad_model_version_id: str = Field(..., description="PatchCore 模型版本 ID")
    detection_confidence: float = Field(default=0.25, ge=0.0, le=1.0)
    efficientad_image_size: int = Field(default=256, ge=64, le=1024)
    efficientad_threshold: float = Field(default=0.99, ge=0.0, le=1.0)
    filter_classifier_model_version_id: str | None = Field(default=None, description="Filter Classifier 模型版本 ID")
    normalizer_model_version_id: str | None = Field(default=None, description="CameraNormalizer stats 模型版本 ID")
    regions: list[CameraRegionCreate] = Field(default_factory=list, description="区域级 PatchCore 配置")


class CameraConfigUpdate(BaseModel):
    camera_id: str | None = Field(default=None, max_length=128)
    efficientad_model_version_id: str | None = Field(default=None)
    detection_confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    efficientad_image_size: int | None = Field(default=None, ge=64, le=1024)
    efficientad_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    filter_classifier_model_version_id: str | None = Field(default=None)
    normalizer_model_version_id: str | None = Field(default=None, description="CameraNormalizer stats 模型版本 ID")
    regions: list[CameraRegionCreate] | None = Field(default=None, description="区域级 PatchCore 配置；传入时整体替换")


class CameraConfigResponse(BaseModel):
    id: str
    camera_id: str
    seat_model_id: str
    efficientad_model_version_id: str | None
    filter_classifier_model_version_id: str | None
    normalizer_model_version_id: str | None = None
    detection_confidence: float
    efficientad_image_size: int
    efficientad_threshold: float
    regions: list[CameraRegionResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Options (dropdown) ──

class CameraOption(BaseModel):
    camera_id: str


class SeatModelOption(BaseModel):
    seat_model_id: str
    display_name: str
    cameras: list[CameraOption] = []
